File: IPWorks Samples/WebDAV Client/java/webdav.java
/*
 * IPWorks 2022 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import java.io.*;
import java.nio.charset.StandardCharsets;

import ipworks.*;

public class webdav extends ConsoleDemo {

    static Webdav webdav = new Webdav();
    static String command;               // user's command
    static String[] argument;            // arguments to the user's command

    public static void main(String[] args) {

        if (args.length < 2) {

            System.out.println("usage: webdav [options] provider username password");
            System.out.println("Options");
            System.out.println("  ");
            System.out.println("  username  the username to login");
            System.out.println("  password  the password to login");
            System.out.println("\r\nExample: webdav username password");

        } else {
            try {
                webdav.addWebdavEventListener(new DefaultWebdavEventListener() {
                    public void SSLServerAuthentication(WebdavSSLServerAuthenticationEvent e) {
                        e.accept = true; //this will trust all certificates and it is not recommended for production use
                    }

                    public void connected(WebdavConnectedEvent e) {
                        System.out.println("Server connected");
                    }

                    public void connectionStatus(WebdavConnectionStatusEvent e) {
                        System.out.println("Status code " + e.statusCode + ": " + e.description);
                    }

                    public void disconnected(WebdavDisconnectedEvent e) {
                        System.out.println("Server disconnected");
                    }
                    // Resource will be empty if data is being posted to the server
                    public void transfer(WebdavTransferEvent e) {
                        System.out.println("Resource being received from server (in full text): \n" +
                                           "========================================= \n" + new String(e.text, StandardCharsets.UTF_8));

                    }

                });

                System.out.println("******************************************************************************");
                System.out.println("* This demo shows how to use the WebDAV component to make directories, move  *");
                System.out.println("* resources, delete resource, and to get resources.                          *");
                System.out.println("******************************************************************************\n");

                System.out.println("Type \"?\" for a list of commands.");

                while (true) {

                    try {
                        System.out.print("> ");
                        command = input();
                        argument = command.split("\\s");

                        if (argument[0].equals("?")) {
                            System.out.println("1) Make Directory");
                            System.out.println("2) Move Resource");
                            System.out.println("3) Get Resource");
                            System.out.println("4) Delete Resource");
                            System.out.println("5) Put Resource");
                            System.out.println("?) This help menu.");
                            System.out.println("Q) Quit.");

                        } else if (argument[0].equals("1")) { // Make Directory
                            webdav.reset();
                            webdav.setUser(args[args.length - 2]);
                            webdav.setPassword(args[args.length - 1]);
                            String server = prompt("Name server where you wish to create a directory (ex. localhost:443)");
                            String dir = prompt("Name directory to create (ex. directory/folder)");
                            webdav.makeDirectory(server + "/" + dir);


                        } else if (argument[0].equals("2")) {  // Move Resource
                            webdav.reset();
                            webdav.setUser(args[args.length - 2]);
                            webdav.setPassword(args[args.length - 1]);
                            String server = prompt("Name server (ex. localhost:443)");
                            String src = prompt("Name source of the resource (ex. myoldfolder/myfile.txt)");
                            String dest = prompt("Name destination of the resource (ex. mynewfolder/myfile.txt)");
                            webdav.moveResource(server + "/" + src, "/" + dest);


                        } else if (argument[0].equals("3")) { // Get Resource
                            webdav.reset();
                            webdav.setUser(args[args.length - 2]);
                            webdav.setPassword(args[args.length - 1]);
                            String get = prompt("Name URI of resource you wish to get");
                            webdav.getResource(get);

                        } else if (argument[0].equals("4")) { // Delete Resource
                            webdav.reset();
                            webdav.setUser(args[args.length - 2]);
                            webdav.setPassword(args[args.length - 1]);
                            String del = prompt("Name URI of resource you wish to delete");
                            webdav.deleteResource(del);

                        } else if (argument[0].equals("5")) { // Put Resource on server, can create resource on server or replace old resource
                            webdav.reset();
                            webdav.setUser(args[args.length - 2]);
                            webdav.setPassword(args[args.length - 1]);
                            String file = prompt("Name path of file you wish to put on server");
                            webdav.setLocalFile(file);
                            String put = prompt("Name URI of resource you wish to put on server ");
                            webdav.putResource(put);

                        }  else if (argument[0].toLowerCase().equals("q")) {
                            System.exit(0);
                            return;
                        } else if (argument[0].equals("")) {
                            // Do nothing
                        } else {
                            System.out.println("Bad command / Not implemented in demo.");
                        } // end of command checking
                    } catch (IPWorksException e) {
                        System.out.println(e.getMessage());
                        e.printStackTrace();
                        System.exit(e.getCode());
                        return;
                    }

                }

            } catch (Exception e) {
                displayError(e);
            }

        }
    }

    public static void WriteToFile(String path, String data) {

        try {
            File outputFile = new File(path);
            outputFile.createNewFile();
            FileOutputStream fw = new FileOutputStream(outputFile.getAbsoluteFile());
            fw.write(data.getBytes());
            fw.flush();
            fw.close();
        } catch (IOException ex) {
            System.out.println("Error writing file " + path + ": " + ex.getMessage());
        }
    }
}
class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }

  static String prompt(String label, String punctuation, String defaultVal)
  {
	System.out.print(label + " [" + defaultVal + "] " + punctuation + " ");
	String response = input();
	if(response.equals(""))
		return defaultVal;
	else
		return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof IPWorksException) {
      System.out.print(" (" + ((IPWorksException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }
}

File: IPWorks Samples/WebDAV Client/javascript/webdav.js
/*
 * IPWorks 2022 JavaScript Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
 
const readline = require("readline");
const ipworks = require("@nsoftware/ipworks");

if(!ipworks) {
  console.error("Cannot find ipworks.");
  process.exit(1);
}
let rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

main()
async function main() {

    const webdav1 = new ipworksdav.webdav();
    let argument;            // arguments to the user's command 
    const argv = process.argv;

    if (argv.length < 2) {

        console.log("usage: node webdav.js username password");
        console.log("Options");
        console.log("  ");
        console.log("username  the username to login");
        console.log("password  the password to login");
        console.log("\r\nExample: webdav username password");
        process.exit();
    } else {
        webdav1.setUser(argv[argv.length - 2]);
        webdav1.setPassword(argv[argv.length - 1]);
        webdav1.on('SSLServerAuthentication', function (e) {
            e.accept = true; //this will trust all certificates and it is not recommended for production use
        })
            .on('Connected', function (e) {
                console.log("Server connected")
            })
            .on('ConnectionStatus', function (e) {
                console.log("Status code " + e.statusCode + ": " + e.description)
            })
            .on('Disconnected', function (e) {
                console.log("Server disconnected")
            })
            .on('Transfer', function (e) {
                console.log("Resource being received from server (in full text): \n" +
                    "========================================= \n" + e.text)
            })

        console.log("******************************************************************************");
        console.log("* This demo shows how to use the WebDAV component to make directories, move *");
        console.log("* resources, delete resource, and to get resources.                          *");
        console.log("******************************************************************************\n");

        rl.setPrompt("Choose an option:\n1) Make Directory\n2) Move Resource\n3) Get Resource\n4) Delete Resource\n5) Put Resource\nQ) Quit.\n> ")
        rl.prompt()

        rl.on('line', async line => {
            argument = line.split(/\s+/);
            var cmd = argument[0];
            if (cmd == "1") { // Make Directory

                rl.question('Name server where you wish to create a directory (ex. localhost:443): ', (server) => {
                    rl.question("Name directory to create (ex. directory/folder): ", async dir => {
                        await webdav1.makeDirectory(server + "/" + dir).catch(ex => console.log("Error: " + ex ))
                        rl.prompt()
                    })
                })

            } else if (cmd == "2") {  // Move Resource
                await webdav1.reset();
                rl.question('Name server (ex. localhost:443): ', (server) => {
                    rl.question("Name source of the resource (ex. myoldfolder/myfile.txt): ", async src => {
                        rl.question("Name destination of the resource (ex. mynewfolder/myfile.txt): ", async dest => {
                            await webdav1.moveResource(server + "/" + src, "/" + dest).catch(ex => console.log("Error: " + ex ))
                            rl.prompt()
                        })
                    })
                })

            } else if (cmd == "3") { // Get Resource
                await webdav1.reset();

                rl.question('Name URI of resource you wish to get: ', async (get) => {
                    await webdav1.getResource(get).catch(ex => console.log("Error: " + ex ))
                    rl.prompt()
                })

            } else if (cmd == "4") { // Delete Resource
                await webdav1.reset();

                rl.question('Name URI of resource you wish to delete: ', async (del) => {
                    await webdav1.deleteResource(del).catch(ex => console.log("Error: " + ex ))
                    rl.prompt()
                })

            } else if (cmd == "5") { // Put Resource on server, can create resource on server or replace old resource
                await webdav1.reset();

                rl.question('Name path of file you wish to put on server: ', (file) => {
                    webdav1.setLocalFile(file);
                    rl.question("Name URI of resource you wish to put on server: ", async put => {
                        await webdav1.putResource(put).catch(ex => console.log("Error: " + ex ))
                        rl.prompt()
                    })
                })

            } else if (cmd == "q") {
                process.exit(0);
            } else if (cmd == "") {
                rl.prompt()
            } else {
                console.log("Bad command / Not implemented in demo.");
                rl.prompt()
            } // end of command checking
        })
    }
}

function prompt(promptName, label, punctuation, defaultVal)
{
  lastPrompt = promptName;
  lastDefault = defaultVal;
  process.stdout.write(`${label} [${defaultVal}] ${punctuation} `);
}
